tests/unit/test_charm.py

```diff
@@ -0,0 +1,145 @@
+import unittest
+from glob import glob
+from unittest.mock import MagicMock, Mock, patch
+
+import lightkube
+from lightkube import codecs
+from lightkube.core.exceptions import ApiError
+from ops.model import ActiveStatus, BlockedStatus
+from ops.testing import Harness
+
+from charm import TrainingOperatorCharm
+
+class _FakeResponse:
+    """Used to fake an httpx response during testing only."""
+
+    def __init__(self, code):
+        self.code = code
+
+    def json(self):
+        if self.code == 409:
+            reason = "AlreadyExists"
+        else: reason = ""
+        return {"apiVersion": 1, "code": self.code, "message": "broken", "reason":reason}
+
+class _FakeApiError(ApiError):
+    """Used to simulate an ApiError during testing."""
+
+    def __init__(self, code=400):
+        super().__init__(response=_FakeResponse(code))
+
+@patch("lightkube.core.client.GenericSyncClient", Mock)
+class TestCharm(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(TrainingOperatorCharm)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    # Event handlers
+    @patch("charm.TrainingOperatorCharm._create_crds")
+    @patch("charm.TrainingOperatorCharm._create_resource")
+    def test_install_event(self, create_auth, create_crd):
+        self.harness.charm.on.install.emit()
+
+        # Ensure create_auth and create_crds
+        # are called once
+        create_auth.assert_called_once()
+        create_crd.assert_called_once()
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating auth resources
+        create_auth.side_effect = _FakeApiError()
+        self.harness.charm.on.install.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus(f"Creating/patching resources failed with code {create_auth.side_effect.response.code}."))
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating CRDs
+        create_crd.side_effect = _FakeApiError()
+        self.harness.charm.on.install.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus(f"Creating/patching resources failed with code {create_crd.side_effect.response.code}."))
```

Comment:
This may just be a style thing, but this feels like three separate tests to me, one for default `install` behaviour and then two when errors aren't handled properly.  Those feel like things to independently test as they can independently break

Reply:
A concrete example of why I'd break this up too is that I think the third test is not actually testing the `create_crd` error catch.  I believe the `self.hardness.charm.on.install.emit()` that is intended to test `create_crd` is actually failing because `create_auth` is mocked and still raising errors (it gets checked before `create_crd`).  A way to prove this would be to put different error codes in the different side effects, but personally I'd break them up into separate tests

Reply:
You are right, the last assertion is not right, definitely will change that. This may be also a good place to use subTest(), I'll experiment with it.

Reply:
Agree with @ca-scribner, This should be 3 different tests.

Reply:
Yeah, maybe not three, but two: one testing on_install basic stuff, like `assert_called_once()`, and the other one testing `BlockedStatus` for each time a diff creation method fails. I decided to go with unittest.subTest() for the latter. Let me know what you think of this approach.